File: AlerteTCL/Views/DataSourceErrorsSheet.swift
import SwiftUI

struct DataSourceErrorsSheet: View {
    @ObservedObject var viewModel: LiveVehiclesViewModel
    @ObservedObject var alertViewModel: AlertViewModel
    @Environment(\.dismiss) private var dismiss
    
    @State private var isRetrying = false
    
    var body: some View {
        NavigationStack {
            List {
                Section {
                    if let error = alertViewModel.error {
                        ErrorRow(
                            title: "Alertes TCL",
                            icon: "exclamationmark.triangle.fill",
                            error: error,
                            isRetrying: isRetrying
                        ) {
                            retryAlerts()
                        }
                    }
                    
                    if let error = viewModel.error {
                        ErrorRow(
                            title: "Véhicules en temps réel",
                            icon: "bus.fill",
                            error: error,
                            isRetrying: isRetrying
                        ) {
                            retryVehicles()
                        }
                    }
                } header: {
                    Text("Sources de données en erreur")
                } footer: {
                    let h = Calendar.current.component(.hour, from: Date())
                    Text(h >= 22 || h < 6
                        ? "Il est probable que Grand Lyon coupe ses serveurs la nuit. Pas d'inquiétude, tout reviendra demain matin !"
                        : "Ces erreurs sont généralement temporaires. Réessayez dans quelques instants."
                    )
                    .font(.footnote)
                    .foregroundStyle(.secondary)
                }
                
                Section {
                    Button {
                        retryAll()
                    } label: {
                        HStack {
                            Image(systemName: "arrow.clockwise")
                            Text("Réessayer toutes les sources")
                            
                            if isRetrying {
                                Spacer()
                                ProgressView()
                                    .scaleEffect(0.8)
                            }
                        }
                    }
                    .disabled(isRetrying)
                }
            }
            .navigationTitle("Erreurs de chargement")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .confirmationAction) {
                    Button("Fermer") {
                        dismiss()
                    }
                }
            }
        }
    }
    
    private func retryAlerts() {
        guard !isRetrying else { return }
        isRetrying = true
        
        Task(priority: .userInitiated) {
            await alertViewModel.loadAlerts()
            await MainActor.run {
                isRetrying = false
            }
        }
    }
    
    private func retryVehicles() {
        guard !isRetrying else { return }
        isRetrying = true
        
        Task(priority: .userInitiated) {
            await viewModel.loadVehicles()
            await MainActor.run {
                isRetrying = false
            }
        }
    }
    
    private func retryAll() {
        guard !isRetrying else { return }
        isRetrying = true
        
        Task(priority: .userInitiated) {
            async let alertsTask: () = alertViewModel.loadAlerts()
            async let vehiclesTask: () = viewModel.loadVehicles()
            _ = await (alertsTask, vehiclesTask)
            
            await MainActor.run {
                isRetrying = false
            }
        }
    }
}

struct ErrorRow: View {
    let title: String
    let icon: String
    let error: String
    let isRetrying: Bool
    let onRetry: () -> Void
    
    var body: some View {
        HStack(spacing: 12) {
            Image(systemName: icon)
                .font(.system(size: 20))
                .foregroundStyle(.orange)
                .frame(width: 32)
            
            VStack(alignment: .leading, spacing: 4) {
                Text(title)
                    .font(.system(size: 15, weight: .medium))
                
                Text(simplifiedError)
                    .font(.system(size: 13))
                    .foregroundStyle(.secondary)
                    .lineLimit(2)
            }
            
            Spacer()
            
            Button {
                onRetry()
            } label: {
                if isRetrying {
                    ProgressView()
                        .scaleEffect(0.8)
                } else {
                    Image(systemName: "arrow.clockwise.circle.fill")
                        .font(.system(size: 24))
                        .foregroundStyle(.blue)
                }
            }
            .buttonStyle(.plain)
            .disabled(isRetrying)
        }
        .padding(.vertical, 4)
    }
    
    private var simplifiedError: String {
        let hour = Calendar.current.component(.hour, from: Date())
        if hour >= 22 || hour < 6 {
            return "Les serveurs Grand Lyon se reposent la nuit — réessayez après 6h"
        }
        if error.contains("504") || error.contains("Timeout") || error.contains("timeout") || error.contains("timed out") {
            return "Le serveur prend son temps... réessayez dans quelques instants"
        } else if error.contains("500") || error.contains("502") || error.contains("503") {
            return "Serveur en maintenance, revenez bientôt"
        } else if error.contains("connection") || error.contains("network") || error.contains("Network") {
            return "Vérifiez votre connexion internet"
        } else if error.contains("401") || error.contains("403") {
            return "Erreur d'authentification avec l'API Grand Lyon"
        } else {
            return "Données temporairement indisponibles"
        }
    }
}

#Preview {
    DataSourceErrorsSheet(
        viewModel: LiveVehiclesViewModel(),
        alertViewModel: AlertViewModel()
    )
}

File: androidApp/src/main/kotlin/com/alertetcl/android/ui/live/DataSourceErrorsSheet.kt
package com.alertetcl.android.ui.live

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.DirectionsBus
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.alertetcl.shared.viewmodels.AlertsViewModel
import com.alertetcl.shared.viewmodels.LiveVehiclesViewModel
import kotlinx.coroutines.launch

@Composable
fun DataSourceErrorsSheet(
    vehiclesError: String?,
    alertsError: String?,
    onRetryVehicles: suspend () -> Unit,
    onRetryAlerts: suspend () -> Unit,
    onDismiss: () -> Unit
) {
    val scope = rememberCoroutineScope()
    var isRetrying by remember { mutableStateOf(false) }

    Column(modifier = Modifier.padding(16.dp).verticalScroll(rememberScrollState())) {
        // Header
        Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(Icons.Filled.Warning, null, tint = MaterialTheme.colorScheme.tertiary, modifier = Modifier.size(24.dp))
            Spacer(Modifier.width(10.dp))
            Text("Erreurs de chargement", fontWeight = FontWeight.Bold, fontSize = 18.sp)
        }
        Spacer(Modifier.height(4.dp))
        Text(
            run {
                val h = java.util.Calendar.getInstance().get(java.util.Calendar.HOUR_OF_DAY)
                if (h >= 22 || h < 6)
                    "Il est probable que Grand Lyon coupe ses serveurs la nuit. Pas d'inquiétude, tout reviendra demain matin !"
                else
                    "Ces erreurs sont généralement temporaires. Réessayez dans quelques instants."
            },
            style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant
        )
        Spacer(Modifier.height(16.dp))
        HorizontalDivider()
        Spacer(Modifier.height(12.dp))

        if (alertsError != null) {
            ErrorRow(
                icon = Icons.Filled.Warning,
                iconTint = MaterialTheme.colorScheme.error,
                title = "Alertes TCL",
                error = alertsError,
                isRetrying = isRetrying
            ) {
                scope.launch {
                    isRetrying = true
                    onRetryAlerts()
                    isRetrying = false
                }
            }
            Spacer(Modifier.height(8.dp))
        }

        if (vehiclesError != null) {
            ErrorRow(
                icon = Icons.Filled.DirectionsBus,
                iconTint = MaterialTheme.colorScheme.primary,
                title = "Véhicules en temps réel",
                error = vehiclesError,
                isRetrying = isRetrying
            ) {
                scope.launch {
                    isRetrying = true
                    onRetryVehicles()
                    isRetrying = false
                }
            }
            Spacer(Modifier.height(8.dp))
        }

        Spacer(Modifier.height(8.dp))
        HorizontalDivider()
        Spacer(Modifier.height(12.dp))

        // Retry all
        Button(
            onClick = {
                scope.launch {
                    isRetrying = true
                    if (vehiclesError != null) onRetryVehicles()
                    if (alertsError != null) onRetryAlerts()
                    isRetrying = false
                }
            },
            enabled = !isRetrying,
            modifier = Modifier.fillMaxWidth(),
        ) {
            if (isRetrying) {
                CircularProgressIndicator(modifier = Modifier.size(16.dp), color = MaterialTheme.colorScheme.onPrimary, strokeWidth = 2.dp)
                Spacer(Modifier.width(8.dp))
            } else {
                Icon(Icons.Filled.Refresh, null, modifier = Modifier.size(16.dp))
                Spacer(Modifier.width(8.dp))
            }
            Text("Réessayer toutes les sources")
        }
        Spacer(Modifier.height(8.dp))
        TextButton(onClick = onDismiss, modifier = Modifier.fillMaxWidth()) {
            Text("Fermer")
        }
        Spacer(Modifier.height(16.dp))
    }
}

@Composable
private fun ErrorRow(
    icon: ImageVector,
    iconTint: Color,
    title: String,
    error: String,
    isRetrying: Boolean,
    onRetry: () -> Unit
) {
    Surface(color = MaterialTheme.colorScheme.errorContainer, tonalElevation = 0.dp, modifier = Modifier.fillMaxWidth()) {
        Column(modifier = Modifier.padding(12.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(icon, null, tint = iconTint, modifier = Modifier.size(18.dp))
                Spacer(Modifier.width(8.dp))
                Text(title, fontWeight = FontWeight.SemiBold, style = MaterialTheme.typography.bodyMedium)
            }
            Spacer(Modifier.height(4.dp))
            Text(error, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onErrorContainer)
            Spacer(Modifier.height(8.dp))
            TextButton(
                onClick = onRetry,
                enabled = !isRetrying
            ) {
                if (isRetrying) {
                    CircularProgressIndicator(modifier = Modifier.size(12.dp), strokeWidth = 1.5.dp)
                    Spacer(Modifier.width(6.dp))
                }
                Text("Réessayer", style = MaterialTheme.typography.bodySmall)
            }
        }
    }
}
